r/tests/testthat/test-embedding.R

```diff
@@ -0,0 +1,37 @@
+mock_req <- function() {
+  req <- list(
+    body = list(
+      type = "umap",
+      config = list(minimumDistance = 0.1, distanceMetric = "cosine")
+    )
+  )
+}
+
+mock_scdata <- function() {
+  data("pbmc_small", package = "SeuratObject", envir = environment())
+  pbmc_small$cells_id <- 0:(ncol(pbmc_small) - 1)
+  pbmc_small@misc$gene_annotations <- data.frame(
+    input = paste0("ENSG", seq_len(nrow(pbmc_small))),
+    name = row.names(pbmc_small),
+    row.names = paste0("ENSG", seq_len(nrow(pbmc_small)))
+  )
+
+  pbmc_small <- Seurat::RunPCA(pbmc_small, npcs = 5)
+  pbmc_small$samples <- rep("Sample1", 80)
+  pbmc_small@misc$numPCs <- 5
+  return(pbmc_small)
+}
+
+test_that("PCA embedding works", {
```

Comment:
are these really the only test cases for embedding? `embedding.R` is 85 lines long, there must be other cases to test for

Reply:
some example cases to test for:

- when there is `numPCs` in `data@misc` and when there is no `numPCs`
- when there is `active.reduction` in `data@misc` and when there is no `active.reduction`
- when there are some wrong values for `active.reduction` and `numPCs`
- when there is no `misc` in `data`
- this line 44 in `embedding.R`: `df_embedding <- Embeddings(data, reduction = type)[, 1:2]` the `[, 1:2]` bit looks very hacky, you can test what happens if `Embeddings` returns something else. You can also go back to the code and try to make it more durable.
- when `type` is not defined or none
- is the data ordered by cell ids in ascending form (I am looking at lines 68-74 in `embedding.R`)
- what if there were no cells in the data? How will lines 68-74 in `embedding.R` behave?

Reply:
Embedding tests are missing because we were waiting on some different issues. I will create a ticket to add what we can, but embedding can't be reproduced in the worker, so it's difficult to test bc the result will vary.